packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -20,64 +23,390 @@ const double _kHandleSize = 22.0;
+        child: Text(localizations.pasteButtonLabel),
+        onPressed: () {
+          setState(() {
+            _overflowOpen = false;
+          });
+          widget.handlePaste();
+        },
+      ));
+    }
+    if (widget.handleSelectAll != null) {
+      _itemKeys.add(GlobalKey());
+      items.add(FlatButton(
+        key: _itemKeys[_itemKeys.length - 1],
+        child: Text(localizations.selectAllButtonLabel),
+        onPressed: () {
+          setState(() {
+            _overflowOpen = false;
+          });
+          widget.handleSelectAll();
+        },
+      ));
+    }
 
     // If there is no option available, build an empty widget.
     if (items.isEmpty) {
       return Container(width: 0.0, height: 0.0);
     }
 
-    return Material(
-      elevation: 1.0,
-      child: Container(
-        height: _kToolbarHeight,
-        child: Row(mainAxisSize: MainAxisSize.min, children: items),
+    // If _itemsInFirstMenu hasn't been calculated yet, render offstage for one
+    // frame for measurement.
+    if (_menuContentWidth == null) {
+      return Offstage(
+        child: _TextSelectionToolbarContainer(
+          key: _containerKey,
+          child: _TextSelectionToolbarContent(
+            items: items,
+            showMoreButton: true,
+            moreButtonKey: _moreButtonKey,
+          ),
+        ),
+      );
+    }
+
+    assert(_indexWhereOverflows != null);
+    final int itemsInFirstMenu = _indexWhereOverflows == -1
+      ? _itemKeys.length
+      : _indexWhereOverflows;
+
+    if (_overflowOpen) {
+      return _TextSelectionToolbarContainer(
+        key: _containerKey,
+        width: _menuContentWidth,
+        child: _TextSelectionToolbarContentOverflow(
+          isAbove: widget.isAbove,
+          items: items.sublist(itemsInFirstMenu, items.length),
+          onBackPressed: () {
+            setState(() {
+              _overflowOpen = false;
+            });
+          },
+        ),
+      );
+    }
+
+    return _TextSelectionToolbarContainer(
+      key: _containerKey,
+      width: _menuContentWidth,
+      child: _TextSelectionToolbarContent(
+        items: items.sublist(0, itemsInFirstMenu),
+        showMoreButton: itemsInFirstMenu < items.length,
+        moreButtonKey: _moreButtonKey,
+        onMorePressed: () {
+          setState(() {
+            _overflowOpen = true;
+          });
+        },
+      ),
+    );
+  }
+}
+
+class _TextSelectionToolbarContainer extends StatefulWidget {
+  const _TextSelectionToolbarContainer({
+    Key key,
+    this.width,
+    this.child,
+  }) : super(key: key);
+
+  final Widget child;
+  final double width;
+
+  @override
+  _TextSelectionToolbarContainerState createState() => _TextSelectionToolbarContainerState();
+}
+
+class _TextSelectionToolbarContainerState extends State<_TextSelectionToolbarContainer> with TickerProviderStateMixin {
+  @override
+  Widget build(BuildContext context) {
+    return Container(
```

Comment:
I just throwing out idea, can we have a RenderObjectElement that can rebuild again if it is not happy about its rendering object? A new RenderObject that take a callback and call the callback after it has been laid out, the Element can use the callback to rebuild if it is not happy about it maybe by marking it dirty or something.

Here you can use a sizedBox instead of container so that it never overflow and will return the width it need

Reply:
Pinging @goderbauer in case you know whether this is possible.

What I would need is to be able to build one widget tree, lay it out, and at that time be able to scrap the widget tree and build and lay out a new one.  The reason is that the widget tree depends on the size of the children to determine if they go into the overflow menu or not.

I took a look at RenderObjectElement and how its used in LayoutBuilder as an example, and I didn't see how this would be possible, but I could be missing it.  For example, [where it calls the `builder` method](https://github.com/flutter/flutter/blob/4896c85ede72a7c9571fb534d21d8570a73697ba/packages/flutter/lib/src/widgets/layout_builder.dart#L100) (and I imagine I could call one of two builder methods or something like that), the layout phase hasn't happened and I wouldn't know the size of the children.  In [performLayout](https://github.com/flutter/flutter/blob/4896c85ede72a7c9571fb534d21d8570a73697ba/packages/flutter/lib/src/widgets/layout_builder.dart#L240) during the layout phase, it doesn't seem possible to build a different child.

Hopefully someone knows for sure either way.

Reply:
I don't think we should make it particularly easy to delay things by a frame as this can easily lead to janky UI.

If you would like to do this all in the same frame you probably could with a custom RenderObject similarly to how we do it with lazy (sliver) lists:

I'd imagine a Toolbar widget that takes two builders:
* The first builder takes a BuildContext and an index. It returns a widget that should be shown in the toolbar.
* The second builder also takes a BuildContext and an index. It returns a widget for the overflow menu with all the items remaining starting at index.

The toolbar widget is now backed by a custom RenderObject that in its layout calls the first builder for index 0 and lays out that child. Now you have its dimensions. If there's room for more, it will call the builder for index 1 and lays that out. And so on.

Once it runs out of space, it would call the second builder to obtain and layout the overflow menu.

Reply:
@goderbauer @chunhtai I was finally able to get this working using RenderBox, needs re-review please!

I left a two questions in Github comments below.  Also, I noticed that several tests are now failing because menu buttons that are laid out but not painted are still detected by `findsNothing`.  This is similarly causing problems with hit detection, which I solved by placing these buttons offscreen.  Is there a way to stop the child from appearing in the tree altogether after calling `child.layout`?